test/run-drun/issue-4323.mo

```diff
@@ -0,0 +1,10 @@
+import Prim = "mo:⛔";
+// let-like annotation
+actor Main : actor { beep : () -> async () } {
+
+  public func beep() : async () {};
+
+  public func ohoh() : async () {Prim.debugPrint "ohoh"}; // still exported, despite not in declared interface!
```

Comment:
Not sure if this is a bug or a feature. What is happening with expressions `module {public foo() {}} : { /* empty */ }`

Reply:
Hmm, type ascription does seem to hide public stuff (contrary to what you mention in #4323) :
``` Motoko
> if true {object Foo {public func foo(){}}; Foo : {}};
stdin:4.1-4.53: warning [M0081], this if has type
  Any
because branches have inconsistent types,
true produces
  {}   👈
false produces
  ()
{foo = func} : Any
```

Reply:
Yes, that's true `<exp> : <typ>` does restrict the type as you say. But in class declarations, the type annotation on the body does not, which I find odd.

```
actor {

    class C() : object {} // this ascription only checks conformance, doesn't restrict
    {
      public func get () {};
    };

    let c = C();
    c.get(); 

}
```
https://play.motoko.org/?tag=1113634911

Reply:
More evidence:

```
actor {

    class C() : object {
      f: () -> Any // note weaker return type
    } // this ascripton only checks conformance, doesn't restrict
    {
      public func get () {};
      public func f() : Nat { 0 };
    };

    let c = C();
    c.get();
    let _ : Nat = c.f();

}
```

https://play.motoko.org/?tag=1445447127

Here. `f` is exported at its actual, not the weaker, specified, type.

Reply:
Yeah, this is probably something we should fix soon.

Reply:
The point of only checking conformance is that this ought to work like an `implements` clause, which is what you normally expect and care about for classes and objects. For hiding you already have public/private in Motoko, so no point in having a duplicate mechanism for that. One way in which Motoko is not ML. :)

So I agree that object/actor/module literals should get an annotation, but that should also be a conformance check only.

Reply:
Isn't that a bit dangerous though? People might expect hiding, given the behaviour of type annotations, with the same colon syntax, on expressions, but then be bitten when, e.g. third party nevertheless successfully call private methods or see extra result info...

Reply:
> One way in which Motoko is not ML. :)

You say that like it's a good thing ;->

Reply:
I added a test run-drun/issue-4323-class.mo that shows the current desugaring is inconistent with the way we type classes (where the this reference has the full type, not the annotated type).

One (ugly) fix would be to change the desugaring to insert a dummy expression that checks the type of `this`, without tripping over the definedness check.
```
import Prim = "mo:⛔";
// let-like annotation
actor This  {
  ignore func () : actor { beep : () -> async () }  { This };
  public func beep() : async () {};

  public func ohoh() : async () {Prim.debugPrint "ohoh"}; // still exported, despite not in declared interface!

  public func check() : async ()  { await This.ohoh() }; // check no curtailment this

}
//CALL ingress ohoh "DIDL\x00\x00"
```
The other would be to extend the syntax of object literals to carry an optional type annotation an use it appropriately in the type checker.

Reply:
> People might expect hiding

I believe that would mostly be ML/FP-biased people. More people probably expect "inheritance" in this case, given the precedence of a colon for that in C++ and others. And "structurally typed interface inheritance" is what they get.

I agree it's not perfect, but like other things in Motoko a compromise with mainstream expectations and "familiarity".

(I also fear that sealing semantics are not a good fit for actors, where static type annotations do not actually hide public entry points, which can still be used from the outside. So such a semantics may even produce a wrong sense of encapsulation.)

> You say that like it's a good thing ;→

Neither, just "different". :)